### ISSUE TITLE
retire Mambaforge and replace it with Miniforge3

### ISSUE THREAD

**Issue by valeriupredoi** [readthedocs/readthedocs.org#11690]
Hey folks, in the RTD [build tools for Python](https://docs.readthedocs.io/en/stable/config-file/v2.html#build-tools-python) all I can see is Mambaforge `mambaforge-latest` - but Mambaforge is being sunsetted, and eventually, retired soon, you folks gonna offer us a Miniforge3 container soon? Cheers muchly :beer:

**Comment by stsewd** [comment]
From the repository https://github.com/conda-forge/miniforge

> As of July 2024, Mambaforge is deprecated. We suggest users switch to Miniforge3 immediately. These installers will be retired from new releases after January 2025. To assist in the migration, we will be introducing rollowing brownouts to the latest Mambaforge installer. Installers up to version 24.5.0-1 will not have any brownouts. 24.5.0-1 will include a warning message. Installers 2024.5.0-2 and later will have the following brownout schedule:
> 
>     The installer will refuse to proceed every two weeks in October
>     The installer will refuse to proceed every ten days in November
>     The installer will refuse to proceed every five days in December
>     The installer will refuse to proceed in 2025+
> 
> Previous information:
> 
> With the [release](https://github.com/conda-forge/miniforge/releases/tag/23.3.1-0) of Miniforge3-23.3.1-0, that incorporated the changes in https://github.com/conda-forge/miniforge/pull/277, the packages and configuration of Mambaforge and Miniforge3 are now identical. The only difference between the two is the name of the installer and, subsequently, the default installation directory.
> 
> We recommend switching to Miniforge3 immediately. These installers will be retired in January 2025. To assist in the migration to Miniforge3 for CI users, we've stopped the latest Mambaforge (24.5+) installer from proceeding with following schedule
> 
>     Every two weeks in October
>     Every ten days in November
>     Every five days in December
>     Never in 2025

**Comment by valeriupredoi** [comment]
@stsewd indeed - some very nasty "brownouts" that we experienced first hand yesterday, hence me scrambling to get us rid of Mambaforge :grin: Many thanks, mate!

**Comment by humitos** [comment]
@valeriupredoi thanks for letting us know about it.

I suppose we should:

- notify all the projects using `miniconda3` and `mambaforge` warning them they are deprecated
- write a blog post explaning this situation to share it with our users
- add `miniforge3` as an option for `build.tools.python`

I quickly check if `miniforge3` is supported on `asdf` and it seems it is:

```
$ asdf list all python | grep miniforge3 | tail
miniforge3-23.1.0-1
miniforge3-23.1.0-2
miniforge3-23.1.0-3
miniforge3-23.1.0-4
miniforge3-23.3.0-0
miniforge3-23.3.1-0
miniforge3-23.3.1-1
miniforge3-23.10.0-0
miniforge3-23.11.0-0
miniforge3-24.3.0-0
```

Unfortunately, the work required for this is not trivial and there is a lot of things to do. I'm not sure how we will prioritize it.

**Comment by valeriupredoi** [comment]
hi @humitos, good luck with the work, and do let me know if I can help in any way! Just be aware of those nasty "brownouts" hitting more frequently in November, and then going full bazooka mode in December; we may switch back to good old Miniconda, while you folks are working on it :beer:

**Comment by humitos** [comment]
I have one question here: is `miniconda3` the same as `miniforge3`? if they are the same, should we only keep support for `miniconda3`? if they are different, what are the differences?

**Comment by duncanmmacleod** [comment]
> I have one question here: is `miniconda3` the same as `miniforge3`? if they are the same, should we only keep support for `miniconda3`? if they are different, what are the differences?

@humitos, no, they are quite different. 

`miniconda3` is the minimal Conda distribution provided by Anaconda Inc that includes `conda` and its dependencies (and maybe a few other things) from the `defaults` channel.

`miniforge3` is the minimal Conda distributino provided by Conda-forge that includes `conda` and its dependencies from the `conda-forge` channel, preconfigured to use _only_ the `conda-forge` channel (notably, not `defaults`).

**Comment by humitos** [comment]
OK, I understand they are the same with the difference where they install packages from: `defaults` or `conda-forge` channels. I also understand that most of our users are using an `environment.yaml` file defining the `channels` key --which in the end it doesn't matter if the underlying application is `miniconda3` or `miniforge3`.

If I'm correct here, I think that adding/keeping support for only one of those is 👍🏼 and either one should be practically the same.

**Comment by duncanmmacleod** [comment]
@humitos the only caveat to your statement comes when users use an `environment.yaml` and _don't_ specify the `channels` key. In that case the install will use whatever channels are configured by the conda distribution, at which point the behaviour will diverge. Sorry if that was so obvious as to not need stating.

**Comment by ericholscher** [comment]
Is it possible to use `build.jobs` to run the proper commands, and then we just don't add support for the new version here? I'd like to avoid the churn, and it seems like folks have been successfully using `uv` with just `build.jobs.install` without us having to support anything?

**Comment by humitos** [comment]
Yeah, I think people could use `build.jobs.pre_create_environment` to install the version of Conda/Mamba they want doing the following:

```yaml
build:
  jobs:
    pre_create_environment:
      - asdf install python miniforge3-24.11.2-1
      - asdf global python miniforge3-24.11.2-1
```

I just checked that and it seems there is an issue in the miniforge3 installer, tho. It fails when installing via `asdf` with ` EnvironmentLocationNotFound: Not a conda environment: /home/docs/checkouts/readthedocs.org/user_builds/test-builds/conda/miniforge3 `, but if you check the logs after that there is nothing relevant on those logs: https://app.readthedocs.org/projects/test-builds/builds/27005828/

So, in any case, before moving in any direction here we should understand why we can't install miniforge3 using asdf first.

**Comment by valeriupredoi** [comment]
@humitos @duncanmmacleod cheers, folks! Heads up though, that the `miniconda-latest` as available from RTD https://docs.readthedocs.com/platform/stable/config-file/v2.html#conda does not support mamba2 which makes a lot of the heftier environments not solve anymore, does RTD support micromamba? I've not seen it listed, so am guessing it's not `asdf`-able 🍻

**Comment by humitos** [comment]
I was able to print some log in https://app.readthedocs.org/projects/test-builds/builds/28007668/. It seems `miniforge3` is installed successfully, but it seems it fails at post processing in `asdf`. I'm not sure yet, but unfortunately we cannot move forward here without understanding what's going on and why we can't install it in Read the Docs.

**Comment by humitos** [comment]
I was able to track this down. We need to `unset` the following environment variables to make `miniforge3` to install properly:

- CONDA_ENVS_PATH
- CONDA_DEFAULT_ENV


This is the command I ran:

```
unset CONDA_ENVS_PATH ; unset CONDA_DEFAULT_ENV ; asdf install python miniforge3-25.1.1-2
asdf global python miniforge3-25.1.1-2
```

Here is the build details page: https://app.readthedocs.org/projects/test-builds/builds/28007879/

With that information, I think we may be able to move forward here.
